### ISSUE TITLE
before-shutdown triggers are fired multiple times

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2692]
|<img alt="yaubi's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| yaubi reported|
|-|-|
|Trac ID|trac#2692|
|Type|defect|
|Created|2007-06-06 14:34:11Z|

Using twisted svn, python 2.4 & 2.5

This program:
```
#!/usr/bin/python
import os, signal, time
from twisted.internet import reactor, defer
def terminate():
    print "terminating ..."
    d = defer.Deferred()
    # do things that lasts long (eg. database, network, etc.)
    reactor.callLater(1, d.callback, None)
    return d
pid = os.getpid()
if os.fork():
    # parent
    reactor.addSystemEventTrigger('before', 'shutdown', terminate)
    reactor.run()
else:
    time.sleep(0.5)
    os.kill(pid, signal.SIGTERM)
    time.sleep(0.5)
    os.kill(pid, signal.SIGTERM)
```

produces the following output:
```
terminating...
terminating...
```

The terminate() method is ran twice, which is not what is excepted. This may also be reproduced by calling reactor.stop() twice (see 2nd test attached)

Proposed patch is attached below. It involves yet-another reactor attribute to discard subsequent signals. It also raises a specialized exception ReactorAlreadyStopped if reactor.stop() is manually called multiple times.

I tryied hard to provide unit tests with this ticket, but unfortunately didn't succeed as it involves reactor.stop (and I'm also not so used to unit tests yet, which is a shame, I know)

See also:
 * [http://twistedmatrix.com/trac/ticket/1485]
 * [http://twistedmatrix.com/trac/ticket/2265]
 * [http://twistedmatrix.com/trac/ticket/2509#comment:7]

Attachments:

* [test1.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d91/d91afe852abb1bd6931c4e534b568e24eb7ff9a9/47c0d18bce28d555845cc8c7eaf10d4b76638051.py) (721 bytes) - added by yaubi on 2007-06-06 14:35:02Z - 1st test, with signals
* [test2.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d91/d91afe852abb1bd6931c4e534b568e24eb7ff9a9/2a41da11c987266d7c39b70a05e3cc3cba268c18.py) (2889 bytes) - added by yaubi on 2007-06-06 14:36:19Z - 2nd test, with reactor.stop
* [patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d91/d91afe852abb1bd6931c4e534b568e24eb7ff9a9/d75b3b528276d7a9f30a04b8bccaf74e1d61f67a) (2549 bytes) - added by yaubi on 2007-06-06 14:38:11Z - Proposed patch

<details><summary>Searchable metadata</summary>

```
trac-id__2692 2692
type__defect defect
reporter__yaubi yaubi
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1181140451000000 1181140451000000
changetime__1193393298000000 1193393298000000
version__None None
owner__ 
cc__yaubi cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks.  I think I know how to write a unit test for this.  I've been making some related fixes for [#2509](https://github.com/twisted/twisted/issues/2509).  I'll probably fix this issue as part of that ticket, too.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve set status to `closed`|
|-|-|

That has been fixed with [#2509](https://github.com/twisted/twisted/issues/2509).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
